### ISSUE TITLE
Enable plugin priority

### ISSUE THREAD

**Issue by giggio** [Flow-Launcher/Flow.Launcher#274]
**Describe the solution you'd like**

I'd like plugins to have a settable priority. When there is more than one plugin offering results for an action keyword, the results should be sorted according to that priority.

**Describe alternatives you've considered**

I can't find any.

**Additional context**

I'd like to have the Window Walker plugin to show results before files, links, directories etc. For example, if I <kbd>Alt</kbd>+<kbd>Space</kbd> and write `Edge` and there is already a Microsoft Edge running, it should show before the Microsoft Edge link or executable.

**Comment by taooceros** [comment]
I like this idea, and I want to have it before but almost forget. I will take a try on it later.
Thank you for suggestion!

**Comment by jjw24** [comment]
so to clarify, this request is where:
- results ordering is based on the priority you set for each of the plugin?
- will frequently used result be higher despite the priority you set?
- this ordering is for ordering each plugin's results not the results given from a plugin right?

**Comment by giggio** [comment]
I believe frequently used options should also have a priority number. And results show in priority order we set on config.
And yes, ordering is per plug-in.
So, for example, I'd set:
1 for Window Walker
2 for frequently used
3 for everything
Etc

If I type "Edge" and I have 2 Edge windows open, they show up first, then if I've used a document named Edge that I've chosen before it would show up, then results for Everything.

**Comment by taooceros** [comment]
> so to clarify, this request is where:
> 
> * results ordering is based on the priority you set for each of the plugin?
> * will frequently used result be higher despite the priority you set?
> * this is for ordering each plugin's results not the results given from within a plugin right?

Hmmm, what do you mean by frequently used result? We haven't implemented it I think #19 ..... If you mean the scoring bump, I believe there are two way to do each.

1. Simply give each plugin priority with a score bump.
2. use orderby for the priority list and thenby for the scoring.

**Comment by taooceros** [comment]
@giggio you can take a try on the pr.
https://ci.appveyor.com/api/buildjobs/qxy09wxucjejhv9e/artifacts/Output%2FPackages%2FFlow-Launcher-v1.6.0.zip

This version may require dotnet core 3.1 to run.

**Comment by jjw24** [comment]
here is the one without needing to download dotnet core 3.1 to run https://ci.appveyor.com/api/buildjobs/jliy0igl2v3s6wgj/artifacts/Output%2FPackages%2FFlow-Launcher-v1.6.0.exe

@giggio Let us know if this is good, the feature is pretty much ready to go

**Comment by giggio** [comment]
@jjw24 I'm getting an error when it launches:

````
Flow Launcher version: 1.6.0
OS Version: Microsoft Windows NT 10.0.19042.0
IntPtr Length: 8
x64: True

Python Path: pythonw.exe
Date: 01/07/2021 11:21:52
Exception:
System.NullReferenceException: Object reference not set to an instance of an object.
   at Flow.Launcher.MainWindow.InitProgressbarAnimation() in C:\projects\flow-launcher\Flow.Launcher\MainWindow.xaml.cs:line 164
   at Flow.Launcher.MainWindow.OnLoaded(Object sender, RoutedEventArgs _) in C:\projects\flow-launcher\Flow.Launcher\MainWindow.xaml.cs:line 66
   at System.Windows.RoutedEventHandlerInfo.InvokeHandler(Object target, RoutedEventArgs routedEventArgs)
   at System.Windows.EventRoute.InvokeHandlersImpl(Object source, RoutedEventArgs args, Boolean reRaised)
   at System.Windows.UIElement.RaiseEventImpl(DependencyObject sender, RoutedEventArgs args)
   at System.Windows.UIElement.RaiseEvent(RoutedEventArgs e)
   at System.Windows.BroadcastEventHelper.BroadcastEvent(DependencyObject root, RoutedEvent routedEvent)
   at System.Windows.BroadcastEventHelper.BroadcastLoadedEvent(Object root)
   at MS.Internal.LoadedOrUnloadedOperation.DoWork()
   at System.Windows.Media.MediaContext.FireLoadedPendingCallbacks()
   at System.Windows.Media.MediaContext.FireInvokeOnRenderCallbacks()
   at System.Windows.Media.MediaContext.RenderMessageHandlerCore(Object resizedCompositionTarget)
   at System.Windows.Media.MediaContext.RenderMessageHandler(Object resizedCompositionTarget)
   at System.Windows.Media.MediaContext.Resize(ICompositionTarget resizedCompositionTarget)
   at System.Windows.Interop.HwndTarget.OnResize()
   at System.Windows.Interop.HwndTarget.HandleMessage(WindowMessage msg, IntPtr wparam, IntPtr lparam)
   at System.Windows.Interop.HwndSource.HwndTargetFilterMessage(IntPtr hwnd, Int32 msg, IntPtr wParam, IntPtr lParam, Boolean& handled)
   at MS.Win32.HwndWrapper.WndProc(IntPtr hwnd, Int32 msg, IntPtr wParam, IntPtr lParam, Boolean& handled)
   at MS.Win32.HwndSubclass.DispatcherCallbackOperation(Object o)
   at System.Windows.Threading.ExceptionWrapper.InternalRealCall(Delegate callback, Object args, Int32 numArgs)
   at System.Windows.Threading.ExceptionWrapper.TryCatchWhen(Object source, Delegate callback, Object args, Int32 numArgs, Delegate catchHandler)
````

**Comment by jjw24** [comment]
fixed:
https://ci.appveyor.com/api/buildjobs/5lwjsshsxhc7dy9u/artifacts/Output%2FPackages%2FFlow-Launcher-v1.7.0.exe

**Comment by jjw24** [comment]
@giggio 
fixed, please give it a spin:
https://ci.appveyor.com/api/buildjobs/5lwjsshsxhc7dy9u/artifacts/Output%2FPackages%2FFlow-Launcher-v1.7.0.exe

**Comment by giggio** [comment]
@jjw24 This version runs!

But I'm not seeing the priorities working. I set the `Program` plugin to priority `1`, `Window Walker` as 2, and the rest as `0`. I'm using `Everything` for my files, and `Explorer` is disabled.

This is what happens when I search for `Sublime`, the Window Walker match is the second:

![image](https://user-images.githubusercontent.com/334958/104196585-cea6a080-53d8-11eb-8138-f26880fa923a.png)

This is what happens when I search for `Process`, while I have `Process Explorer` running, the Window Walker match only shows as third result:

![image](https://user-images.githubusercontent.com/334958/104196659-e54cf780-53d8-11eb-9e60-aa7fe4b444c4.png)

This is what happens when I search for `Edge`, while I have the Edge browser running, the Window Walker match only shows as fifth result:

![image](https://user-images.githubusercontent.com/334958/104196973-4bd21580-53d9-11eb-9d6b-080dca7343fd.png)

It seems executables are ranking higher, it seems that from the `Program plugin.

I tried restarting Flow, it doesn't help.

**Comment by taooceros** [comment]
Well, I think you can set the priority of window walker more higher (for example 3 or 4), because the scoring of the fuzzy search of window walker seems much lower than the other plugin. For detail, I only set one priority to increase 50 in scoring. As a result, please make it higher for now. I think this should be changed higher actually.




------------------&nbsp;原始邮件&nbsp;------------------
发件人: "Giovanni Bassi"<notifications@github.com&gt;; 
发送时间: 2021年1月11日(星期一) 晚上10:53
收件人: "Flow-Launcher/Flow.Launcher"<Flow.Launcher@noreply.github.com&gt;; 
抄送: "天煞"<1569418337@qq.com&gt;; "Assign"<assign@noreply.github.com&gt;; 
主题: Re: [Flow-Launcher/Flow.Launcher] Enable plugin priority (#274)





 
@jjw24 This version runs!
 
But I'm not seeing the priorities working. I set the Program plugin to priority 1, Window Walker as 2, and the rest as 0. I'm using Everything for my files, and Explorer is disabled.
 
This is what happens when I search for Sublime, the Window Walker match is the second:
 

 
This is what happens when I search for Process, while I have Process Explorer running, the Window Walker match only shows as third result:
 

 
This is what happens when I search for Edge, while I have the Edge browser running, the Window Walker match only shows as fifth result:
 

 
It seems executables are ranking higher, it seems that from the `Program plugin.
 
I tried restarting Flow, it doesn't help.

**Comment by giggio** [comment]
Yes, changing Window Walker to priority 4 worked, all 3 cases are working fine now!
If it is kept this way I suggest you add that note to the docs, other people will surely encounter a similar issue. Or maybe some help text on the settings screen.

**Comment by taooceros** [comment]
Yeah, will do. Thank you! I just wondering whether someone may want a more specific control of scoring bump (like 50 in window walker may be higher than 70 in program, but lower than 90), or simply make once result higher than other will be good enough.




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "Flow-Launcher/Flow.Launcher"                                                                                    <notifications@github.com&gt;;
发送时间:&nbsp;2021年1月11日(星期一) 晚上11:04
收件人:&nbsp;"Flow-Launcher/Flow.Launcher"<Flow.Launcher@noreply.github.com&gt;;
抄送:&nbsp;"天煞"<1569418337@qq.com&gt;;"Assign"<assign@noreply.github.com&gt;;
主题:&nbsp;Re: [Flow-Launcher/Flow.Launcher] Enable plugin priority (#274)





 
Yes, changing Window Walker to priority 4 worked, all 3 cases are working fine now!
 If it is kept this way I suggest you add that note to the docs, other people will surely encounter a similar issue. Or maybe some help text on the settings screen.

**Comment by giggio** [comment]
I don't understand how the scoring works in Flow, so I can't help you on that.

**Comment by taooceros** [comment]
Anyway, thank you so much for testing!😀
This should be released in 1.7.0!


------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "Flow-Launcher/Flow.Launcher"                                                                                    <notifications@github.com&gt;;
发送时间:&nbsp;2021年1月11日(星期一) 晚上11:13
收件人:&nbsp;"Flow-Launcher/Flow.Launcher"<Flow.Launcher@noreply.github.com&gt;;
抄送:&nbsp;"天煞"<1569418337@qq.com&gt;;"Assign"<assign@noreply.github.com&gt;;
主题:&nbsp;Re: [Flow-Launcher/Flow.Launcher] Enable plugin priority (#274)





 
I don't understand how the scoring works in Flow, so I can't help you on that.
